sql/core/src/main/scala/org/apache/spark/sql/execution/joins/RowIterator.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import java.util.NoSuchElementException
+
+import org.apache.spark.sql.catalyst.InternalRow
+
+private[sql] abstract class RowIterator {
+  def advanceNext(): Boolean
+  def getRow: InternalRow
+  def toScala: Iterator[InternalRow] = new RowIteratorToScala(this)
+}
+
+object RowIterator {
+  def fromScala(scalaIter: Iterator[InternalRow]): RowIterator = {
+    scalaIter match {
+      case wrappedRowIter: RowIteratorToScala => wrappedRowIter.rowIter
+      case _ => new RowIteratorFromScala(scalaIter)
+    }
+  }
+}
+
+private final class RowIteratorToScala(val rowIter: RowIterator) extends Iterator[InternalRow] {
+  private [this] var _hasNext: Boolean = rowIter.advanceNext()
+  override def hasNext: Boolean = _hasNext
+  override def next(): InternalRow = {
+    if (!_hasNext) throw new NoSuchElementException
+    val row: InternalRow = rowIter.getRow.copy()
```

Comment:
I'm going to try to remove this copy and will add some assertions that will fail in the cases where the copy would have been necessary. This will make it easy to spot whether it's necessary as of this release. If it does turn out to be a problem and looks like it will take a long time to fix, then I'll defer the removal of the copy to a smaller followup patch.